src/metabase/api/dashboard.clj

```diff
@@ -787,37 +787,53 @@
     (keyword (name type))
     :=))
 
+(defn- reverse-join [join]
+  {:rhs (:lhs join)
+   :lhs (:rhs join)})
+
 (mu/defn ^:private param->fields
-  [{:keys [mappings] :as param} :- mbql.s/Parameter]
+  [{:keys [mappings] :as param} :- mbql.s/Parameter top-level?]
   (for [{:keys [target] {:keys [card]} :dashcard} mappings
         :let  [[_ dimension] (->> (mbql.normalize/normalize-tokens target :ignore-path)
                                   (mbql.u/check-clause :dimension))]
         :when dimension
-        :let  [ttag      (get-template-tag dimension card)
-               dimension (condp mbql.u/is-clause? dimension
-                           :field        dimension
-                           :template-tag (:dimension ttag)
-                           (log/error "cannot handle this dimension" {:dimension dimension}))
-               field-id  (or
-                          ;; Get the field id from the field-clause if it contains it. This is the common case
-                          ;; for mbql queries.
-                          (mbql.u/match-one dimension [:field (id :guard integer?) _] id)
-                          ;; Attempt to get the field clause from the model metadata corresponding to the field.
-                          ;; This is the common case for native queries in which mappings from original columns
-                          ;; have been performed using model metadata.
-                          (:id (qp.util/field->field-info dimension (:result_metadata card))))]
-        :when field-id]
-    {:field-id field-id
+        :let  [ttag         (get-template-tag dimension card)
+               dimension    (condp mbql.u/is-clause? dimension
+                              :field        dimension
+                              :template-tag (:dimension ttag)
+                              (log/error "cannot handle this dimension" {:dimension dimension}))
+               [id options] (or
+                             ;; Get the field id from the field-clause if it contains it. This is the common case
+                             ;; for mbql queries.
+                             (mbql.u/match-one dimension [:field (id :guard integer?) options] [id options])
```

Comment:
[this](https://github.com/metabase/metabase/blob/efa24940113e26a2141a329bfdd96a74f725d11f/shared/src/metabase/mbql/schema.cljc#L323) is the schema of field options, it could contains stuff like base-type or binning.

we surely don't want this to be used as expression options right?

this was my worry from the previous PR([comment](https://github.com/metabase/metabase/pull/34837#discussion_r1375920142))

Reply:
But we don't use those `options` for anything else except look for `:join-alias` on them, so should be fine? Our constraint gets options from template tag and from param, but not from a field dimension.

Or do you still think we could end up with excessive stuff in Constraint?